rust/worker/src/execution/data/data_chunk.rs

```diff
@@ -0,0 +1,146 @@
+use std::ops::Range;
+use std::sync::Arc;
+
+use crate::types::EmbeddingRecord;
+
+#[derive(Clone, Debug)]
+pub(crate) struct DataChunk {
+    data: Arc<[EmbeddingRecord]>,
+    visibility: Arc<[bool]>,
+}
+
+impl DataChunk {
+    pub fn new(data: Arc<[EmbeddingRecord]>) -> Self {
+        let len = data.len();
+        DataChunk {
+            data,
+            visibility: vec![true; len].into(),
+        }
+    }
+
+    pub fn len(&self) -> usize {
```

Comment:
i think it may make more semantic sense for len() to be the visible_len and to instead have physical_len() which is the size of the data